### PR TITLE
feat(core): 🎸 compatible with async futures that rely on tokio runtime

### PR DESCRIPTION
**Pull request by wjian23**
…ntime

### CODE REVIEW

File: core/Cargo.toml [hunk 1]
```diff
@@ -39,6 +39,15 @@ colored.workspace = true
...
 ribir_dev_helper = {path = "../dev-helper"}
 ribir = { path = "../ribir" }
+tokio = {version = "1.0", features = ["full"]}
 
 [features]
+default = ["tokio-async"]
 png = ["ribir_painter/png"]
+tokio-async = ["tokio"]
+
+[dependencies.tokio]
```
Comment (M-Adoo) [inline core/Cargo.toml:49]: move to workspace cargo.toml

File: core/src/context/app_ctx.rs [hunk 2]
```diff
@@ -443,6 +452,326 @@ impl Drop for AppCtxScopeGuard {
...
+      if !waker.is_null() {
+        unsafe {
+          let _ = Box::from_raw(waker);
+        };
+      }
+    }
+  }
+
+  #[cfg(test)]
+  mod tests {
```
Comment (M-Adoo) [inline core/src/context/app_ctx.rs:723]: tests can merge into the next `tests` mod

File: core/src/context/app_ctx.rs [hunk 3]
```diff
@@ -443,6 +452,326 @@ impl Drop for AppCtxScopeGuard {
...
+        .swap(std::ptr::null_mut(), std::sync::atomic::Ordering::Relaxed);
+      if !waker.is_null() {
+        unsafe {
+          let _ = Box::from_raw(waker);
+        };
+      }
+    }
+  }
+
+  impl<F> Future for BlockingFuture<F>
```
Comment (M-Adoo) [inline core/src/context/app_ctx.rs:616]: seems `FutureExt::remote_handle`  do same thing.

File: core/src/context/app_ctx.rs [hunk 4]
```diff
@@ -443,6 +452,326 @@ impl Drop for AppCtxScopeGuard {
...
+      Fut: Future + Send + 'static,
+      Fut::Output: Send,
+    {
+      LocalWaitFuture {
+        waker: Arc::new(AtomicPtr::new(std::ptr::null_mut())),
+        fut: BlockFuture::Raw(future),
+      }
+    }
+
+    pub fn tokio_runtime() -> &'static tokio::runtime::Runtime {
```
Comment (M-Adoo) [inline core/src/context/app_ctx.rs:473]: should be private

**Review by M-Adoo — APPROVED** [review]

**Review by sologeek — APPROVED** [review]

**Pull request merged**